### PR TITLE
Convert OOP-based code to ECS

### PR DESCRIPTION
**Pull request by utilForever**
This revision includes:
- Convert OOP-based code to ECS (#17)
  - Add library `entt`
  - Convert classes to components using `struct`
  - Refactor related code

### CODE REVIEW

**Comment by circle-oo** [comment]
> This revision includes:
> 
> * Convert OOP-based code to ECS ([Convert OOP-based code to ECS #17](https://github.com/utilForever/PokeMaster/issues/17))
>   
>   * Add library `entt`
>   * Convert classes to components using `struct`
>   * Refactor related code

File: Includes/PokeMaster/Helpers/PokemonHelpers.hpp [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+//! \param registry A registry of the entity-component system.
+//! \param entity A pokemon entity.
+//! \return The level of the pokemon.
+int GetLevel(entt::registry& registry, entt::entity entity);
+
+//! Returns the types of the pokemon.
+//! \param registry A registry of the entity-component system.
+//! \param entity A pokemon entity.
+//! \return The types of the pokemon.
+std::tuple<Type, Type> GetTypes(entt::registry& registry, entt::entity entity);
```
Comment (circle-oo) [inline Includes/PokeMaster/Helpers/PokemonHelpers.hpp:78]: Is `std::tuple<>` better than `std::pair<>`?

File: Includes/PokeMaster/Components/Stats.hpp [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+};
+
+//!
+//! \brief Stats struct.
+//!
+//! A statistic (Japanese: 能力 ability), or stat for short, is an element which
+//! determines certain aspects of battles in the games. Stats may also refer to the
+//! numerical values of each field in regards to individual Pokémon.
+//!
+struct Stats
```
Comment (circle-oo) [inline Includes/PokeMaster/Components/Stats.hpp:62]: How about add simple comment about each values 'base', 'individual' and 'effort'?

File: Includes/PokeMaster/Helpers/PokemonHelpers.hpp (hunk 1 above)
Comment (utilForever) [reply Includes/PokeMaster/Helpers/PokemonHelpers.hpp:78]: `std::tuple<>` is generalized type of `std::pair<>`. `std::pair<>` stores two types only. However, `std::tuple<>` can store arbitrary number of types as you wish.

File: Includes/PokeMaster/Components/Stats.hpp (hunk 2 above)
Comment (utilForever) [reply Includes/PokeMaster/Components/Stats.hpp:62]: Okay. I'll add simple comment for these values.

File: Includes/PokeMaster/Systems/Game.hpp [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+
+    //! Returns the level of the pokemon.
+    //! \param entity A pokemon entity.
+    //! \return The level of the pokemon.
+    int GetPokemonLevel(entt::entity entity);
+
+    //! Returns the types of the pokemon.
+    //! \param entity A pokemon entity.
+    //! \return The types of the pokemon.
+    std::tuple<Type, Type> GetPokemonTypes(entt::entity entity);
```
Comment (circle-oo) [inline Includes/PokeMaster/Systems/Game.hpp:62]: Is `std::tuple` better than `std::pair`?

Comment (utilForever) [reply Includes/PokeMaster/Systems/Game.hpp:62]: Same as above.

File: Includes/PokeMaster/Helpers/PokemonHelpers.hpp [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+void LoadData(entt::registry& registry);
+
+//! Adds a pokemon for the player.
+//! \param registry A registry of the entity-component system.
+//! \param name The name of the pokemon.
+//! \param level The level of the pokemon.
+//! \param individualValues A list of individual values.
+//! \param effortValues A list of effort values.
+//! \param nature The nature of pokemon.
+//! \return An added pokemon entity.
```
Comment (circle-oo) [inline Includes/PokeMaster/Helpers/PokemonHelpers.hpp:45]:
```suggestion
//! Adds a Pokémon for the player.
//! \param registry A registry of the entity-component system.
//! \param name The name of the Pokémon.
//! \param level The level of the Pokémon.
//! \param individualValues A list of individual values.
//! \param effortValues A list of effort values.
//! \param nature The nature of Pokémon.
//! \return An added Pokémon entity.
```

File: Includes/PokeMaster/Helpers/PokemonHelpers.hpp [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+entt::entity Add(entt::registry& registry, std::string_view&& name, int level,
+                 StatStorage individualValues, StatStorage effortValues, Nature nature);
+
+//! Calculates stats of pokemon with base, IV, EV, level and nature.
+//! \param registry A registry of the entity-component system.
+//! \param entity A pokemon entity.
```
Comment (circle-oo) [inline Includes/PokeMaster/Helpers/PokemonHelpers.hpp:51]:
```suggestion
//! Calculates stats of Pokémon with base, IV, EV, level and nature.
//! \param registry A registry of the entity-component system.
//! \param entity A Pokémon entity.
```

File: Includes/PokeMaster/Helpers/PokemonHelpers.hpp [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+void CalculateStats(entt::registry& registry, entt::entity entity);
+
+//! Returns a pokemon entity that matches \p index.
+//! \param registry A registry of the entity-component system.
+//! \param index The pokemon index to find.
+//! \return A pokemon entity that matches \p index.
```
Comment (circle-oo) [inline Includes/PokeMaster/Helpers/PokemonHelpers.hpp:57]:
```suggestion
//! Returns a Pokémon entity that matches \p index.
//! \param registry A registry of the entity-component system.
//! \param index The Pokémon index to find.
//! \return A Pokémon entity that matches \p index.
```

File: Includes/PokeMaster/Helpers/PokemonHelpers.hpp [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+[[nodiscard]] std::optional<entt::entity> FindByIndex(entt::registry& registry,
+                                                      int index);
+
+//! Returns a pokemon entity that matches \p name.
+//! \param registry A registry of the entity-component system.
+//! \param name The pokemon name to find.
+//! \return A pokemon entity that matches \p name.
```
Comment (circle-oo) [inline Includes/PokeMaster/Helpers/PokemonHelpers.hpp:64]:
```suggestion
//! Returns a Pokémon entity that matches \p name.
//! \param registry A registry of the entity-component system.
//! \param name The Pokémon name to find.
//! \return A Pokémon entity that matches \p name.
```

File: Includes/PokeMaster/Helpers/PokemonHelpers.hpp [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+[[nodiscard]] std::optional<entt::entity> FindByName(entt::registry& registry,
+                                                     std::string_view&& name);
+
+//! Returns the level of the pokemon.
+//! \param registry A registry of the entity-component system.
+//! \param entity A pokemon entity.
+//! \return The level of the pokemon.
```
Comment (circle-oo) [inline Includes/PokeMaster/Helpers/PokemonHelpers.hpp:71]:
```suggestion
//! Returns the level of the Pokémon.
//! \param registry A registry of the entity-component system.
//! \param entity A Pokémon entity.
//! \return The level of the Pokémon.
```

File: Includes/PokeMaster/Helpers/PokemonHelpers.hpp [hunk 9]
```diff
@@ -0,0 +1,87 @@
...
+int GetLevel(entt::registry& registry, entt::entity entity);
+
+//! Returns the types of the pokemon.
+//! \param registry A registry of the entity-component system.
+//! \param entity A pokemon entity.
+//! \return The types of the pokemon.
```
Comment (circle-oo) [inline Includes/PokeMaster/Helpers/PokemonHelpers.hpp:77]:
```suggestion
//! Returns the types of the Pokémon.
//! \param registry A registry of the entity-component system.
//! \param entity A Pokémon entity.
//! \return The types of the Pokémon.
```

File: Includes/PokeMaster/Helpers/PokemonHelpers.hpp [hunk 10]
```diff
@@ -0,0 +1,87 @@
...
+std::tuple<Type, Type> GetTypes(entt::registry& registry, entt::entity entity);
+
+//! Returns the stats of the pokemon.
+//! \param registry A registry of the entity-component system.
+//! \param entity A pokemon entity.
+//! \return The stats of the pokemon.
```
Comment (circle-oo) [inline Includes/PokeMaster/Helpers/PokemonHelpers.hpp:83]:
```suggestion
//! Returns the stats of the Pokémon.
//! \param registry A registry of the entity-component system.
//! \param entity A Pokémon entity.
//! \return The stats of the Pokémon.
```

File: Includes/PokeMaster/Systems/Game.hpp [hunk 11]
```diff
@@ -0,0 +1,74 @@
...
+    //! Constructs new game and loads related data.
+    Game();
+
+    //! Adds a pokemon for the player.
+    //! \param name The name of the pokemon.
+    //! \param level The level of the pokemon.
+    //! \param individualValues A list of individual values.
+    //! \param effortValues A list of effort values.
+    //! \param nature The nature of pokemon.
+    //! \return An added pokemon entity.
```
Comment (circle-oo) [inline Includes/PokeMaster/Systems/Game.hpp:39]:
```suggestion
    //! Adds a Pokémon for the player.
    //! \param name The name of the Pokémon.
    //! \param level The level of the Pokémon.
    //! \param individualValues A list of individual values.
    //! \param effortValues A list of effort values.
    //! \param nature The nature of Pokémon.
    //! \return An added Pokémon entity.
```

File: Includes/PokeMaster/Systems/Game.hpp [hunk 12]
```diff
@@ -0,0 +1,74 @@
...
+    entt::entity AddPokemon(std::string_view name, int level,
+                            StatStorage individualValues, StatStorage effortValues,
+                            Nature nature);
+
+    //! Returns a pokemon entity that matches \p index.
+    //! \param index The pokemon index to find.
+    //! \return A pokemon entity that matches \p index.
```
Comment (circle-oo) [inline Includes/PokeMaster/Systems/Game.hpp:46]:
```suggestion
    //! Returns a Pokémon entity that matches \p index.
    //! \param index The Pokémon index to find.
    //! \return A Pokémon entity that matches \p index.
```

File: Includes/PokeMaster/Systems/Game.hpp [hunk 13]
```diff
@@ -0,0 +1,74 @@
...
+    [[nodiscard]] std::optional<entt::entity> FindPokemonByIndex(int index);
+
+    //! Returns a pokemon entity that matches \p name.
+    //! \param name The pokemon name to find.
+    //! \return A pokemon entity that matches \p name.
```
Comment (circle-oo) [inline Includes/PokeMaster/Systems/Game.hpp:51]:
```suggestion
    //! Returns a Pokémon entity that matches \p name.
    //! \param name The Pokémon name to find.
    //! \return A Pokémon entity that matches \p name.
```

File: Includes/PokeMaster/Systems/Game.hpp [hunk 14]
```diff
@@ -0,0 +1,74 @@
...
+    [[nodiscard]] std::optional<entt::entity> FindPokemonByName(std::string_view&& name);
+
+    //! Returns the level of the pokemon.
+    //! \param entity A pokemon entity.
+    //! \return The level of the pokemon.
```
Comment (circle-oo) [inline Includes/PokeMaster/Systems/Game.hpp:56]:
```suggestion
    //! Returns the level of the Pokémon.
    //! \param entity A Pokémon entity.
    //! \return The level of the Pokémon.
```

File: Includes/PokeMaster/Systems/Game.hpp [hunk 15]
```diff
@@ -0,0 +1,74 @@
...
+    int GetPokemonLevel(entt::entity entity);
+
+    //! Returns the types of the pokemon.
+    //! \param entity A pokemon entity.
+    //! \return The types of the pokemon.
```
Comment (circle-oo) [inline Includes/PokeMaster/Systems/Game.hpp:61]:
```suggestion
    //! Returns the types of the Pokémon.
    //! \param entity A Pokémon entity.
    //! \return The types of the Pokémon.
```

File: Includes/PokeMaster/Systems/Game.hpp [hunk 16]
```diff
@@ -0,0 +1,74 @@
...
+    std::tuple<Type, Type> GetPokemonTypes(entt::entity entity);
+
+    //! Returns the stats of the pokemon.
+    //! \param entity A pokemon entity.
+    //! \return The stats of the pokemon.
```
Comment (circle-oo) [inline Includes/PokeMaster/Systems/Game.hpp:66]:
```suggestion
    //! Returns the stats of the Pokémon.
    //! \param entity A Pokémon entity.
    //! \return The stats of the Pokémon.
```

File: Tests/UnitTests/Systems/GameTests.cpp [hunk 17]
```diff
@@ -0,0 +1,127 @@
...
+// personal capacity and are not conveying any rights to any intellectual
+// property of any third parties.
+
+#include "doctest_proxy.hpp"
+
+#include <PokeMaster/Systems/Game.hpp>
+
+using namespace PokeMaster;
+
+TEST_CASE("[Game] - FindPokemonByIndex")
```
Comment (circle-oo) [inline Tests/UnitTests/Systems/GameTests.cpp:13]:
```suggestion
TEST_CASE("[Game] - FindPokémonByIndex")

```

File: Tests/UnitTests/Systems/GameTests.cpp [hunk 18]
```diff
@@ -0,0 +1,127 @@
...
+        CHECK_EQ(pokemon.has_value(), true);
+    }
+
+    {
+        auto pokemon = game.FindPokemonByIndex(-1);
+        CHECK_EQ(pokemon.has_value(), false);
+    }
+}
+
+TEST_CASE("[Game] - FindPokemonByName")
```
Comment (circle-oo) [inline Tests/UnitTests/Systems/GameTests.cpp:28]:
```suggestion
TEST_CASE("[Game] - FindPokémonByName")

```

File: Tests/UnitTests/Systems/GameTests.cpp [hunk 19]
```diff
@@ -0,0 +1,127 @@
...
+        CHECK_EQ(pokemon.has_value(), true);
+    }
+
+    {
+        auto pokemon = game.FindPokemonByName("test");
+        CHECK_EQ(pokemon.has_value(), false);
+    }
+}
+
+TEST_CASE("[Game] - GetPokemonLevel")
```
Comment (circle-oo) [inline Tests/UnitTests/Systems/GameTests.cpp:43]:
```suggestion
TEST_CASE("[Game] - GetPokémonLevel")

```

File: Tests/UnitTests/Systems/GameTests.cpp [hunk 20]
```diff
@@ -0,0 +1,127 @@
...
+    {
+        auto pokemon =
+            game.AddPokemon("dragonair", 24, StatStorage{ 15, 24, 8, 18, 23, 29 },
+                            StatStorage{ 0, 0, 252, 4, 252, 0 }, Nature::HARDY);
+        auto level = game.GetPokemonLevel(pokemon);
+        CHECK_EQ(level, 24);
+    }
+}
+
+TEST_CASE("[Game] - GetPokemonTypes")
```
Comment (circle-oo) [inline Tests/UnitTests/Systems/GameTests.cpp:64]:
```suggestion
TEST_CASE("[Game] - GetPokémonTypes")

```

File: Tests/UnitTests/Systems/GameTests.cpp [hunk 21]
```diff
@@ -0,0 +1,127 @@
...
+        auto pokemon = game.FindPokemonByName("dragonair");
+        CHECK_EQ(pokemon.has_value(), true);
+
+        auto [type1, type2] = game.GetPokemonTypes(pokemon.value());
+        CHECK_EQ(type1, Type::DRAGON);
+        CHECK_EQ(type2, Type::INVALID);
+    }
+}
+
+TEST_CASE("[Game] - GetPokemonStats")
```
Comment (circle-oo) [inline Tests/UnitTests/Systems/GameTests.cpp:87]:
```suggestion
TEST_CASE("[Game] - GetPokémonStats")

```

File: Tests/UnitTests/Systems/GameTests.cpp (hunk 17 above)
Comment (utilForever) [reply Tests/UnitTests/Systems/GameTests.cpp:13]: Don't agree. It's method name.

File: Tests/UnitTests/Systems/GameTests.cpp (hunk 18 above)
Comment (utilForever) [reply Tests/UnitTests/Systems/GameTests.cpp:28]: Same as above.

File: Tests/UnitTests/Systems/GameTests.cpp (hunk 19 above)
Comment (utilForever) [reply Tests/UnitTests/Systems/GameTests.cpp:43]: Same as above.

File: Tests/UnitTests/Systems/GameTests.cpp (hunk 20 above)
Comment (utilForever) [reply Tests/UnitTests/Systems/GameTests.cpp:64]: Same as above.

File: Tests/UnitTests/Systems/GameTests.cpp (hunk 21 above)
Comment (utilForever) [reply Tests/UnitTests/Systems/GameTests.cpp:87]: Same as above.

File: Includes/PokeMaster/Helpers/PokemonHelpers.hpp (hunk 4 above)
Comment (utilForever) [reply Includes/PokeMaster/Helpers/PokemonHelpers.hpp]: Fix in 9f9bef4b7cdf0492978b3d55a1b5188587d8e877.

File: Includes/PokeMaster/Helpers/PokemonHelpers.hpp (hunk 5 above)
Comment (utilForever) [reply Includes/PokeMaster/Helpers/PokemonHelpers.hpp]: Fixed in 9f9bef4b7cdf0492978b3d55a1b5188587d8e877.

File: Includes/PokeMaster/Helpers/PokemonHelpers.hpp (hunk 6 above)
Comment (utilForever) [reply Includes/PokeMaster/Helpers/PokemonHelpers.hpp]: Fixed in 9f9bef4b7cdf0492978b3d55a1b5188587d8e877.

File: Includes/PokeMaster/Helpers/PokemonHelpers.hpp (hunk 7 above)
Comment (utilForever) [reply Includes/PokeMaster/Helpers/PokemonHelpers.hpp]: Fixed in 9f9bef4b7cdf0492978b3d55a1b5188587d8e877.

File: Includes/PokeMaster/Helpers/PokemonHelpers.hpp (hunk 8 above)
Comment (utilForever) [reply Includes/PokeMaster/Helpers/PokemonHelpers.hpp]: Fixed in 9f9bef4b7cdf0492978b3d55a1b5188587d8e877.

File: Includes/PokeMaster/Helpers/PokemonHelpers.hpp (hunk 9 above)
Comment (utilForever) [reply Includes/PokeMaster/Helpers/PokemonHelpers.hpp]: Fixed in 9f9bef4b7cdf0492978b3d55a1b5188587d8e877.

File: Includes/PokeMaster/Helpers/PokemonHelpers.hpp (hunk 10 above)
Comment (utilForever) [reply Includes/PokeMaster/Helpers/PokemonHelpers.hpp]: Fixed in 9f9bef4b7cdf0492978b3d55a1b5188587d8e877.

File: Includes/PokeMaster/Systems/Game.hpp (hunk 11 above)
Comment (utilForever) [reply Includes/PokeMaster/Systems/Game.hpp]: Fixed in 9f9bef4b7cdf0492978b3d55a1b5188587d8e877.

File: Includes/PokeMaster/Systems/Game.hpp (hunk 12 above)
Comment (utilForever) [reply Includes/PokeMaster/Systems/Game.hpp]: Fixed in 9f9bef4b7cdf0492978b3d55a1b5188587d8e877.

File: Includes/PokeMaster/Systems/Game.hpp (hunk 13 above)
Comment (utilForever) [reply Includes/PokeMaster/Systems/Game.hpp]: Fixed in 9f9bef4b7cdf0492978b3d55a1b5188587d8e877.

File: Includes/PokeMaster/Systems/Game.hpp (hunk 14 above)
Comment (utilForever) [reply Includes/PokeMaster/Systems/Game.hpp]: Fixed in 9f9bef4b7cdf0492978b3d55a1b5188587d8e877.

File: Includes/PokeMaster/Systems/Game.hpp (hunk 15 above)
Comment (utilForever) [reply Includes/PokeMaster/Systems/Game.hpp]: Fixed in 9f9bef4b7cdf0492978b3d55a1b5188587d8e877.

File: Includes/PokeMaster/Systems/Game.hpp (hunk 16 above)
Comment (utilForever) [reply Includes/PokeMaster/Systems/Game.hpp]: Fixed in 9f9bef4b7cdf0492978b3d55a1b5188587d8e877.

**Review by circle-oo — APPROVED** [review]

**Pull request merged**